lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,348 @@
+        assert(llvm::isa_and_nonnull<ConstantOp>(def));
+        inst->setOperand(oldEnablePort, enable);
+        if (def->getUses().empty())
+          opsToErase.insert(def);
+        continue;
+      }
+
+      clockEnablePortIndices[index] = enablePortNum++;
+      gatedClkPorts.push_back(index);
+
+      // Create a name for the new enable port.
+      auto clockIndexStr = std::to_string(index);
+      enablePorts.emplace_back(
+          StringAttr::get(context,
+                          (argNames.empty()
+                               ? "enable_"
+                               : argNames[index].getValue() + "_enable_") +
+                              clockIndexStr),
+          enable);
+    }
+    if (enablePorts.empty())
+      continue;
+    unsigned oldNumInputs = inst->getNumOperands();
+
+    // Now update the instance and all the referenced modules with the new
+    // enable ports.
+    appendInputsToModuleAndInstances(inst, moduleInstances, enablePorts,
+                                     refMod);
+    // Third phase, Once all the modules are updated with the additional enable
+    // ports, duplicate the clock_gate op in the referenced module.
+    auto *block = refMod.getBodyBlock();
+    auto builder = mlir::OpBuilder::atBlockBegin(block);
+    for (auto [index, clkPort] : llvm::enumerate(gatedClkPorts)) {
+      auto clk = block->getArgument(clkPort);
+      assert(isa<ClockType>(clk.getType()));
+      auto enable = block->getArgument(oldNumInputs + index);
+      auto clkGate = builder.create<ClockGateOp>(enable.getLoc(), clk, enable,
+                                                 Value(), hw::InnerSymAttr());
+      // Replace all the original clock users with the gated clock.
+      clk.replaceAllUsesExcept(clkGate.getResult(), clkGate);
+    }
+    // After adding the new clock gates, scan the module for instances that
+    // have a gated clock as input, and continue sinking the clock gates.
+    findInstancesWithGatedClock(refMod);
+  }
+  for (auto *gatedOp : opsWithGatedClock) {
+    LLVM_DEBUG(llvm::dbgs() << "\n Gated clock user: " << gatedOp);
+    // TODO: Handle all possible ops here, to replace the clocked op with the
+    // ungated clock and sink the enable into the op.
+    if (auto compReg = dyn_cast<CompRegOp>(gatedOp)) {
+      Value clk = compReg.getClk();
+      auto gate = dyn_cast_or_null<ClockGateOp>(clk.getDefiningOp());
+      if (!gate)
+        continue;
+      clk = gate.getInput();
+      Value enable = getEnable(gate);
+      OpBuilder builder(context);
+      builder.setInsertionPoint(compReg);
+      auto compEn = builder.create<CompRegClockEnabledOp>(
+          compReg->getLoc(), compReg.getInput(), clk, enable,
+          compReg.getNameAttr(), compReg.getReset(), compReg.getResetValue(),
+          compReg.getPowerOnValue(), compReg.getInnerSymAttr());
+      compReg->replaceAllUsesWith(compEn);
+      opsToErase.insert(compReg);
+    }
```

Comment:
I think this should go to canonicalization if it's always beneficial.

Reply:
Yeah, I thought about that, but as you pointed out, I don't think this is "beneficial" in all cases. This should only happen if the user explicitly enables this transformation. So, the transformation of "sink clock-gates" implies trying to combine the clock gate's enable into the operation that is driven by the gated clock. This might not be desirable in general, and if we add it to the canonicalization we can't disable it in general. 
But maybe this can be added to some op interface. A new op interface that any op can opt-in, and then we can invoke it from here.
My plan is to add the enable to FirRegOp next, similar transformation needs to happen to memory ops also, but its semantics are slightly non-trivial.